pkg/client/bucket/bucket.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	// Client abstracts the API access for Grail Buckets
+	Client struct {
+		url     string
+		client  *rest.Client
+		limiter *concurrency.Limiter
+	}
+)
+
+// NewClient creates a new client to interact with the Grail Bucket API
+func NewClient(url string, client *rest.Client) *Client {
+	return &Client{
+		url:     url,
+		client:  client,
+		limiter: concurrency.NewLimiter(5),
+	}
+}
+
+// Upsert create or updates a given Grail Bucket
+func (c Client) Upsert(ctx context.Context, id string, data []byte) (result Response, err error) {
+	if id == "" {
+		return Response{}, fmt.Errorf("id must be non empty")
+	}
+	c.limiter.ExecuteBlocking(func() {
+		result, err = c.upsert(ctx, id, data)
+	})
+	return
+}
+
+// upsert first tries to create the bucket. If it does not work, we try to update it.
+// This has the advantage that we don't need to try to fetch the object first.
+func (c Client) upsert(ctx context.Context, id string, data []byte) (Response, error) {
+	r, err := c.create(ctx, id, data)
+	if err == nil {
+		log.WithCtxFields(ctx).Debug("created object with ID %q", id)
+		return r, nil
+	}
+
+	log.WithCtxFields(ctx).Debug("failed to create new object with ID %q; trying to update existing: %w", id, err)
+
+	b, err := c.get(ctx, id)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to get object with ID %q: %w", id, err)
+	}
+
+	r, err = c.update(ctx, b, data)
+	log.WithCtxFields(ctx).Debug("updated object with ID %q", id)
+	return r, err
+}
+
+func (c Client) create(ctx context.Context, id string, data []byte) (Response, error) {
+	u, err := url.JoinPath(c.url, endpoint)
+	if err != nil {
+		return Response{}, fmt.Errorf("faild to create sound url: %w", err)
+	}
+
+	err = setBucketName(id, &data)
+	if err != nil {
+		return Response{}, err
+	}
+
+	r, err := c.client.Post(ctx, u, data)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to create object with ID %q: %w", id, err)
+	}
+	if !r.IsSuccess() {
+		return Response{}, rest.NewRespErr(fmt.Sprintf("failed to update object with ID %q (HTTP %d): %s", id, r.StatusCode, string(r.Body)), r).WithRequestInfo(http.MethodPut, u)
+	}
+
+	b, err := unmarshalJSON(r.Body)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to read response: %w", err)
+	}
+
+	return b, nil
+}
+
+func (c Client) update(ctx context.Context, b Response, data []byte) (Response, error) {
+
+	u, err := url.Parse(c.url)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to parse url: %w", err)
+	}
+
+	u.Path, err = url.JoinPath(u.Path, endpoint, b.BucketName)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to join url: %w", err)
+	}
+
+	q := u.Query()
+	q.Add("optimistic-locking-version", strconv.Itoa(b.Version))
+	u.RawQuery = q.Encode()
+
+	var m map[string]any
+	err = json.Unmarshal(data, &m)
+	if err != nil {
+		return Response{}, fmt.Errorf("unable to unmarshal template: %w", err)
+	}
+	m["bucketName"] = b.BucketName
+	m["version"] = b.Version
+	m["status"] = b.Status
```

Comment:
is the status required to be set on update? could we just omit it?

Reply:
It's required. We can't update a bucket to something that it isn't, e.g. 'creating'